docs/experimental/nnx/mnist_tutorial.md

```diff
@@ -0,0 +1,320 @@
+test_ds = test_ds.shuffle(1024) # create shuffled dataset by allocating a buffer size of 1024 to randomly draw elements from
+test_ds = test_ds.batch(batch_size, drop_remainder=True).prefetch(1) # group into batches of batch_size and skip incomplete batch, prefetch the next sample to improve latency
+```
+
+## 3. Define the Network with NNX
+
+Create a convolutional neural network with NNX by subclassing `nnx.Module`.
+
+```{code-cell} ipython3
+from flax.experimental import nnx  # NNX API
+
+class CNN(nnx.Module):
+  """A simple CNN model."""
+
+  def __init__(self, *, rngs: nnx.Rngs):
+    self.conv1 = nnx.Conv(in_features=1, out_features=32, kernel_size=(3, 3), rngs=rngs)
+    self.conv2 = nnx.Conv(in_features=32, out_features=64, kernel_size=(3, 3), rngs=rngs)
+    self.linear1 = nnx.Linear(in_features=3136, out_features=256, rngs=rngs)
+    self.linear2 = nnx.Linear(in_features=256, out_features=10, rngs=rngs)
+
+  def __call__(self, x):
+    x = self.conv1(x)
+    x = nnx.relu(x)
+    x = nnx.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+    x = self.conv2(x)
+    x = nnx.relu(x)
+    x = nnx.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
+    x = x.reshape((x.shape[0], -1))  # flatten
+    x = self.linear1(x)
+    x = nnx.relu(x)
+    x = self.linear2(x)
+    return x
+  
+model = CNN(rngs=nnx.Rngs(0))
+
+print(f'model = {model}'[:500] + '\n...\n')  # print a part of the model
+print(f'{model.conv1.kernel.shape = }') # inspect the shape of the kernel of the first convolutional layer
+```
+
+### Run model
+
+Let's put our model to the test!  We'll perform a forward pass with arbitrary data and print the results.
+
+```{code-cell} ipython3
+:outputId: 2c580f41-bf5d-40ec-f1cf-ab7f319a84da
+
+import jax
+import jax.numpy as jnp  # JAX NumPy
+
+y = model(jnp.ones((1, 28, 28, 1)))
+y
+```
+
+## 4. Define Metrics
+
+To track our model's performance, we'll use the [clu](https://github.com/google/CommonLoopUtils) library. If you haven't already, install it with:
+
+```{code-cell} ipython3
+!pip install -q clu
+```
+
+Let's create a compound metric using clu.metrics.Collection.  This will include both an Accuracy metric for tracking how well our model classifies images, and an Average metric to monitor the average loss over each training epoch.
+
+```{code-cell} ipython3
+from clu import metrics
+from flax import struct   # Flax pytree dataclasses
+
+@struct.dataclass
+class Metrics(metrics.Collection):
+  accuracy: metrics.Accuracy
+  loss: metrics.Average.from_output('loss')
+```
+
+## 5. Create the `TrainState`
+
+In Flax, a common practice is to use a dataclass to encapsulate the training state, including the step number, parameters, and optimizer state. The [`flax.training.train_state.TrainState`](https://flax.readthedocs.io/en/latest/flax.training.html#train-state) class is ideal for basic use cases, simplifying the process by allowing you to pass a single argument to functions like `train_step`.
```

Comment:
Interesting, the fusion of Flax and NNX!

Do we have to use `TrainState` in this example? Does it make sense to make it more self-contained and free of abstractions from non-NNX parts of Flax?

Reply:
The Functional API works 🙂. However, I've been thinking of creating a Module-based `TrainState` that would be even more convenient that the current Pytree API.